api/types/access_request.go

```diff
@@ -185,6 +206,255 @@ func (r *AccessRequestV3) SetSystemAnnotations(annotations map[string][]string)
 	r.Spec.SystemAnnotations = annotations
 }
 
+func (r *AccessRequestV3) GetOriginalRoles() []string {
+	if l := len(r.Spec.RoleThresholdMapping); l == 0 || l == len(r.Spec.Roles) {
+		// rtm is unspecified or original role list is unmodified
+		return r.Spec.Roles
+	}
+
+	// role subselection has been applied.  calculate original roles
+	// by collecting the keys of the rtm.
+	roles := make([]string, 0, len(r.Spec.RoleThresholdMapping))
+	for role := range r.Spec.RoleThresholdMapping {
+		roles = append(roles, role)
+	}
+	sort.Strings(roles)
+	return roles
+}
+
+// getThreshold is a helper for getting a threshold by its stored index.
+func (r *AccessRequestV3) getThreshold(tid uint32) (AccessReviewThreshold, error) {
+	idx := int(tid)
+	if len(r.Spec.Thresholds) <= idx {
+		return AccessReviewThreshold{}, trace.Errorf("threshold index '%d' out of range (this is a bug)", tid)
+	}
+	return r.Spec.Thresholds[idx], nil
+}
+
+// SetThresholds sets the review thresholds.
+func (r *AccessRequestV3) SetThresholds(thresholds []AccessReviewThreshold) {
+	r.Spec.Thresholds = thresholds
+}
+
+// SetRoleThresholdMapping sets the rtm (internal use only).
+func (r *AccessRequestV3) SetRoleThresholdMapping(rtm map[string]ThresholdIndexSets) {
+	r.Spec.RoleThresholdMapping = rtm
+}
+
+// ApplyReview applies an access review.  The supplied parser must be pre-configured to evaluate
+// boolean expressions based on the review and its author.  Prefer using services.ApplyAccessReview
+// rather than calling this method directly.
+func (r *AccessRequestV3) ApplyReview(rev AccessReview, parser predicate.Parser) error {
+	if !rev.State.IsApproved() && !rev.State.IsDenied() {
+		return trace.BadParameter("invalid state proposal: %s (expected approval/denial)", rev.State)
+	}
+
+	// the default theshold should exist. if it does not, the request either is not fully
+	// initialized (i.e. variable expansion has not been run yet) or the request was inserted into
+	// the backend by a teleport instance which does not support the review feature.
+	if len(r.Spec.Thresholds) == 0 {
+		return trace.BadParameter("request is uninitialized or does not support reviews")
+	}
+
+	for _, existingReview := range r.Spec.Reviews {
+		if existingReview.Author == rev.Author {
+			return trace.AccessDenied("user %q has already reviewed this request", rev.Author)
+		}
+	}
```

Comment:
Our old model allowed changing request state after it was already approved, and that ended up being a bad decision in retrospect.  The access request system grants privileges by allowing the user to provision certificates with additional roles.  If a request is changed from approved to some other non-approved state, it gives the false impression that the privileges were never assumed.  Once a request has been in the approved state for even a few milliseconds, you cannot discount the possibility that the user has assumed elevated privileges.  Since our caching layer is fairly asynchronous, re-checking the request state on every user action isn't a great solution to this.  Its better to have the request continue to reflect the fact that the user likely did successfully escalate privilege.

In terms of cancelling reviews before the request's state-transition occurs, I'm not fundamentally opposed.  I don't see any pressing need to add it on the initial release of this feature though.